spec/index.bs

```diff
@@ -605,31 +603,57 @@ interface GPUProgrammablePassEncoder : GPUObjectBase {
     void insertDebugMarker(DOMString markerLabel);
 };
 
-interface GPURenderPassEncoder : GPUProgrammablePassEncoder {
+interface GPURenderEncoderBase : GPUProgrammablePassEncoder {
```

Comment:
For D3D12, bundles were intended to provide reusability of lightweight command list fragments. However, the D3D team's experience so far suggests they are rarely a performance win, and in many cases a pessimisation.  Their programming model is too restricted to map well to typical app usage patterns, and they inhibit optimizations that drivers would otherwise perform. The [Root Signature documentation](https://docs.microsoft.com/en-us/windows/desktop/direct3d12/root-signature-version-1-1) talks about this in the overview section. For this reason they have not been widely adopted, and the D3D team does not recommend their use in WebGPU. We lose similar optimizations if we reuse command lists.  

I understand that, under the covers, the API can coalesce the web developer's bundles and command lists into larger ones.  I appreciate efforts minimize the amount of Javascript that has to run.  At the same time, I worry about optimizations like this limiting the API's ability to be as low level as possible.

Reply:
Can you elaborate how they could inhibit driver optimizations? In the worst case, can't the driver just treat it as a replay into the command buffer?

I think it's not necessary, and possibly not useful, to use D3D bundles to implement WebGPU bundles. If we (e.g. Dawn) replay the bundle ourselves, it should definitely be identical from the driver's perspective.

IIRC our bundles allow commands that aren't allowed in D3D bundles or Vulkan secondary command buffers, anyway.

Reply:
FWIW, I've been arguing for us to limit the bundle functions to the intersection of all Vulkan SCB/D3D12 Bundle/Metal ICB low-level APIs.

@RafaelCintron thank you for the info!
> The Root Signature documentation talks about this in the overview section.

Could you quote the most relevant bits describing why bundles inhibit driver optimizations? As far as I understand, D3D12 bundles don't affect the descriptors in the heap, and they can't change the current heap, so I'm not completely clear on what the problem is here.

Reply:
Thanks @RafaelCintron for sharing the experience from D3D. To add on what @kainino0x said, Bundle in WebGPU will also be used to record the inside of render passes in parallel. Until recently D3D12 only had `OMSetRenderTarget` and no render passes, it didn't need to make the distinction between command buffers and render bundles. WebGPU has strict render passes so we need to have this two level scheme to allow applications to record in parallel at a granularity smaller than the main render pass.

Reply:
@kvark , the section I was referring to was the part where it talks about descriptors and data being static. If you mark things as `DATA_STATIC`, for example, that means you're promising D3D12 that both the descriptors and the data pointed to by those descriptors will never change until the command list and bundles referencing them have finished executing for the last time.  Drivers use this information to optimize shaders or promote descriptors to root descriptors behind your back at command list recording time.  In order to change the descriptor or data pointer by the descriptor between different submissions of the same command list, you must use one or more of the "volatile" flags talked about in the document.  

In D3D11, everything is sequential and multi-queues are not directly exposed to developers.  Hence, drivers had all of the information they needed to perform the optimizations on your behalf.  So by adding the volatile flags, not only are you forging the optimizations, you'll also get worse performance than you would if you used D3D11 with the same workload.

Reply:
While we're on the subject of volatile descriptors and D3D12, one part of the document that is relevant to WebGPU is the following section:

> When descriptors are static there is another change in behavior that requires the DESCRIPTORS_VOLATILE flag to be set. Out of bounds accesses to any Buffer views (as opposed to Texture1D/2D/3D/Cube views) are invalid and produce undefined results, including possible device reset, rather than returning default values for reads or dropping writes. The purpose for removing the ability for applications to depend on hardware out of bounds access checking is to allow drivers to choose to promote static descriptor accesses to root descriptor accesses if they deem that more efficient. Root descriptors don’t support any out of bounds checking.

Newer versions of D3D12 added `DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS` flag.  This flag disallows descriptor promotion described above while still keeping the shader specific optimizations.

Reply:
Thanks for great feedback @RafaelCintron !
At the very minimum, we'll be able to set `DATA_STATIC_WHILE_SET_AT_EXECUTE` mode most of the time, since the resources are bound in scope of a pass, and no transfer operations are allowed there. Is this enough to get at least some optimizations by the driver?

With this proposal API, we can go further and use `DATA_STATIC` on uniform buffers. When we detect that the contents are changed, we can create another descriptor (similar to copy-on-write) to use.

Reply:
With `DATA_STATIC_WHILE_SET_AT_EXECUTE`, `SetXXXXRootDescriptorTable` will invalidate `STATIC_WHILE_SET_AT_EXECUTE` prefetched data cache and shader optimizations.  `DATA_STATIC` descriptors, on the other hand, remain static across `SetXXXXRootDescriptorTable` calls, retaining the prefetch cache and optimized shader bindings. This means you can get reduced GPU perf due to extra cache flushes.

@kvark , I agree it is possible to do `DATA_STATIC` on uniform buffers and use copy-on-write to detect that the contents are changing.  But that brings us closer an API with the overhead OpenGL and D3D11 have today. One of the goals of this API is to put more power into the hands of web developers to make these kinds of optimizations/tradeoffs.

Reply:
@RafaelCintron (trying to clarify, it's not yet completely clear to me)

> SetXXXXRootDescriptorTable will invalidate STATIC_WHILE_SET_AT_EXECUTE prefetched data cache and shader optimizations

Supposing we had a descriptor of this type, and we called `SetXXXXRootDescriptorTable`. Surely, the data prefetched from this descriptor is no longer needed to be kept around. In what scenario would this be a performance concern? The draw calls that had been made with this descriptor supposedly got it in an optimized way (via the driver promoting the data into the root constants). What happens after the descriptor change is another story. I could think of a case where the user re-binds the previously used descriptor, and with DATA_STATIC that means the driver needs to do less work. Is this where we'd get an extra cache flush?

As for shader optimizations, that part I don't understand. I expect the shader to depend on the root signature, not the currently bound descriptors. If the root signature says that this particular descriptor is `STATIC_WHILE_SET_AT_EXECUTE` then as long as this signature is used, I'd expect all shaders to have the access to this descriptor optimized, regardless of any `SetXXXXRootDescriptorTable` calls. Is my understanding wrong here?

> But that brings us closer an API with the overhead OpenGL and D3D11 have today.

Now that I think of it, using `DATA_STATIC` in such a manner seems unfeasible for bundles: copying one descriptor on write wouldn't magically update the recorded bundle :)

Reply:
@kvark , the [Root Signature](https://docs.microsoft.com/en-us/windows/desktop/direct3d12/root-signature-version-1-1) documentation describes the guarantees that come with each flag.  The details of the optimizations are intentionally left vague because they're IHV specific and can change overtime. Generally speaking, the sooner the driver knows data is static, the more time it has to pre-fetch data and do optimizations.  

With `DATA_STATIC`, you, as the developer, are guaranteeing that the data pointed to by descriptors has been initialized by the time a root descriptor or descriptor table referencing the memory has been set on a command list / bundle **during recording**, and the data cannot be changed until the command list / bundle has finished executing for the last time.  This flag is good for cases where all of your uniform buffers and descriptors are locked in and ready to go when you record the command list.

With `DATA_STATIC_WHILE_SET_AT_EXECUTE`, data pointed to by descriptors cannot change starting from when the underlying root descriptor or descriptor table is set on a command list / bundle during execution on the GPU timeline, and ending when subsequent draws/dispatches will no longer reference the data. As a developer, the only guarantees you're making to the driver are about the state of the data **during execution while the descriptor is bound**.  You're free to change the data while you're recording the command list on the CPU.  You're also free to change the data at execution time when the descriptor is not bound. For instance, if you go from descriptor A to B and back to A again, the driver needs to re-figure out optimizations for A the second time A is bound.  

For for applications that do a good job of submitting GPU work in parallel across multiple threads, it isn’t always feasible to synchronize such that all generation of input data happens before recording the command lists which use it.  In that case, it is still beneficial to communicate to the driver that all data will be available by execute time, even if it wasn’t yet available at record time.

> Now that I think of it, using DATA_STATIC in such a manner seems unfeasible for bundles: copying one descriptor on write wouldn't magically update the recorded bundle :)

I agree that if you copy the descriptor on write, you have to re-record command lists/bundles that you use it.

Reply:
Thank you for the detailed clarification!
The conclusions I make from it are:
  - we can use `DATA_STATIC_WHILE_SET_AT_EXECUTE` safely
  - with this mode, the driver can still do the optimization of promoting constants to the root signature. However, it would have to copy the actual constants out when the command buffer  is *executed* as opposed to the *recording* time available for `DATA_STATIC`
  - whether or not the driver would actually do this is unknown, depends on the platform

If I understand this correctly, D3D12 bundles could still be useful for our performance.